File: 783.MinimumDistance BetweenBSTNodes.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var minDiffInBST = function(root) {
    let min = Infinity;
    const stack = [];
    let current = root;
    let previous;
    while(true) {
        if (current) {
            stack.push(current);
            current = current.left;
        } else if (stack.length) {
            const node = stack.pop();
            if (node && previous) {
                min = Math.min(min, Math.abs(previous.val - node.val)); 
            }
            previous = node;
            current = node.right;
        } else {
            break;
    }
    return min;
};

File: Golang/Tree/783.Minimum-distance-between-bst-nodes.go
package main

import "math"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func minDiffInBST(root *TreeNode) int {
	ret := 100000
	var prev *TreeNode
	node := root
	stack := []*TreeNode{}
	for {
		if node != nil {
			stack = append(stack, node)
			node = node.Left
		} else if len(stack) > 0 {
			temp := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			if prev != nil {
				diff := int(math.Abs(float64(prev.Val) - float64(temp.Val)))
				if diff < ret {
					ret = diff
				}
			}
			prev = temp
			node = temp.Right
		} else {
			break
		}
	}
	return ret
}
